pkg/provisioners/gcppubsub/controller/channel/reconcile.go

```diff
@@ -178,20 +219,100 @@ func (r *reconciler) reconcile(ctx context.Context, c *eventingv1alpha1.Channel)
 		return false, err
 	}
 
-	topic, err := r.createTopic(ctx, c, gcpCreds, r.defaultGcpProject)
+	topic, err := r.createTopic(ctx, plannedPCS, gcpCreds)
 	if err != nil {
 		return false, err
 	}
 
-	err = r.createSubscriptions(ctx, c, gcpCreds, r.defaultGcpProject, topic)
+	err = r.syncSubscriptions(ctx, plannedPCS, gcpCreds, topic, subsToSync)
 	if err != nil {
 		return false, err
 	}
+	// Now that the subs have synced successfully, remove the old ones from the status.
+	plannedPCS.Subscriptions = subsToSync.subsToCreate
+	if err = pubsubutil.SaveRawStatus(ctx, c, plannedPCS); err != nil {
+		return false, err
+	}
 
 	c.Status.MarkProvisioned()
 	return false, nil
 }
 
+// planGcpResources creates the plan for every resource that needs to be created outside of the
+// cluster. Their names are all saved to the returned in the updated status.
+func (r *reconciler) planGcpResources(ctx context.Context, c *eventingv1alpha1.Channel, originalPCS *pubsubutil.GcpPubSubChannelStatus) (persistence, *pubsubutil.GcpPubSubChannelStatus, *syncSubs, error) {
+	persist := noNeedToPersist
+	subsToSync := &syncSubs{
+		subsToCreate: make([]pubsubutil.GcpPubSubSubscriptionStatus, 0),
+		subsToDelete: make([]pubsubutil.GcpPubSubSubscriptionStatus, 0),
+	}
+
+	topicName := originalPCS.Topic
+	if topicName == "" {
+		topicName = generateTopicName(c)
+	}
+	// Everything except the subscriptions.
+	newPCS := &pubsubutil.GcpPubSubChannelStatus{
+		// TODO Allow arguments to set the secret and project.
+		Secret:     r.defaultSecret,
+		SecretKey:  r.defaultSecretKey,
+		GCPProject: r.defaultGcpProject,
+		Topic:      topicName,
+	}
+	if !equality.Semantic.DeepEqual(originalPCS.Secret, newPCS.Secret) || originalPCS.SecretKey != newPCS.SecretKey || originalPCS.GCPProject != newPCS.GCPProject || originalPCS.Topic != newPCS.Topic {
+		persist = persistStatus
+	}
+
+	existingSubs := make(map[types.UID]pubsubutil.GcpPubSubSubscriptionStatus, len(originalPCS.Subscriptions))
+	for _, existingSub := range originalPCS.Subscriptions {
+		// I don't think this can ever happen, but let's just be sure.
```

Comment:
Because we use the `Ref.UID` as the key to correlate Subscriptions between `spec` and `status`, it needs to be present. If it is not present, then the logic about how to correlate Subscriptions in the `spec` and in the `status` doesn't work. `Ref` is a pointer, so can be `nil` or the `Ref.UID` may be the empty string, which means we shouldn't reconcile it. 

This shouldn't happen because Subscriptions are only ever written if the UID is present. But in case it does happen, we don't want a panic due to dereferencing a `nil` pointer. Nor do we want to accidentally correlate Subscriptions between `spec` and `status` if they both have the empty string as the UID.

Reply:
Done.